File: src/engine/adapters/mysql-family/definition.ts
import type { ConstraintMeta, ObjectDefinition } from '../../../shared/domain/definition';
import { encodePath, type NodePath } from '../../../shared/domain/tree';
import { AdapterError } from '../errors';
import { stripOneTrailingSemicolon } from '../sql-text';
import type { QueryExecutor } from './catalog';
import { quoteIdent } from './read';

export type RelationLikeKind = 'table' | 'view';

interface TableTypeRow {
  TABLE_TYPE: string;
}

interface TableConstraintRow {
  name: string;
  type: string; // 'PRIMARY KEY' | 'UNIQUE' | 'FOREIGN KEY' | 'CHECK'
  check_clause: string | null;
}

interface KeyColumnRow {
  name: string;
  col: string;
  ref_table: string | null;
  ref_col: string | null;
}

const CONSTRAINT_TYPE: Record<string, ConstraintMeta['type']> = {
  'PRIMARY KEY': 'primaryKey',
  UNIQUE: 'unique',
  'FOREIGN KEY': 'foreignKey',
  CHECK: 'check',
};

// MariaDB has no `pg_get_constraintdef`-style builtin (P19 D11) — the key-column list and the
// FK's referenced table/columns are composed from information_schema itself, not invented SQL
// syntax; a CHECK constraint's clause is the engine's own CHECK_CONSTRAINTS.CHECK_CLAUSE text,
// rendered verbatim like everything else in this file.
async function listConstraints(
  exec: QueryExecutor,
  database: string,
  tableName: string,
): Promise<ConstraintMeta[]> {
  const constraints = await exec<TableConstraintRow>(
    `SELECT tc.CONSTRAINT_NAME AS name, tc.CONSTRAINT_TYPE AS type, cc.CHECK_CLAUSE AS check_clause
     FROM information_schema.TABLE_CONSTRAINTS tc
     LEFT JOIN information_schema.CHECK_CONSTRAINTS cc
       ON cc.CONSTRAINT_SCHEMA = tc.CONSTRAINT_SCHEMA AND cc.CONSTRAINT_NAME = tc.CONSTRAINT_NAME
     WHERE tc.TABLE_SCHEMA = ? AND tc.TABLE_NAME = ?
     ORDER BY FIELD(tc.CONSTRAINT_TYPE, 'PRIMARY KEY', 'UNIQUE', 'CHECK', 'FOREIGN KEY'),
              tc.CONSTRAINT_NAME`,
    [database, tableName],
  );
  if (constraints.length === 0) return [];

  const keyColumns = await exec<KeyColumnRow>(
    `SELECT CONSTRAINT_NAME AS name, COLUMN_NAME AS col,
            REFERENCED_TABLE_NAME AS ref_table, REFERENCED_COLUMN_NAME AS ref_col
     FROM information_schema.KEY_COLUMN_USAGE
     WHERE TABLE_SCHEMA = ? AND TABLE_NAME = ?
     ORDER BY CONSTRAINT_NAME, ORDINAL_POSITION`,
    [database, tableName],
  );
  const columnsByConstraint = new Map<string, KeyColumnRow[]>();
  for (const row of keyColumns) {
    const list = columnsByConstraint.get(row.name) ?? [];
    list.push(row);
    columnsByConstraint.set(row.name, list);
  }

  return constraints.map((c): ConstraintMeta => {
    const type = CONSTRAINT_TYPE[c.type] ?? 'check';
    if (type === 'check') {
      return { name: c.name, type, definition: c.check_clause ?? '' };
    }
    const cols = columnsByConstraint.get(c.name) ?? [];
    const columnList = `(${cols.map((r) => r.col).join(', ')})`;
    if (type === 'foreignKey') {
      const refTable = cols[0]?.ref_table ?? '';
      const refColumnList = `(${cols.map((r) => r.ref_col ?? '').join(', ')})`;
      return {
        name: c.name,
        type,
        definition: `${columnList} REFERENCES ${refTable} ${refColumnList}`,
      };
    }
    return { name: c.name, type, definition: columnList };
  });
}

/**
 * Passes `SHOW CREATE …` through verbatim — MariaDB is asked, never composed. The one identifier
 * that must reach the statement text goes through `quoteIdent`, resolved from
 * `information_schema` (with bound parameters) in this same op — Adapter rule 7.
 */
export async function buildDefinition(
  exec: QueryExecutor,
  segments: NodePath['segments'],
  database: string,
  object: { kind: RelationLikeKind; name: string },
): Promise<ObjectDefinition> {
  const [resolved] = await exec<TableTypeRow>(
    `SELECT TABLE_TYPE FROM information_schema.TABLES WHERE TABLE_SCHEMA = ? AND TABLE_NAME = ?`,
    [database, object.name],
  );
  if (!resolved) {
    throw new AdapterError('E_NOT_FOUND', `relation "${database}"."${object.name}" not found`);
  }

  const qualified = `${quoteIdent(database)}.${quoteIdent(object.name)}`;
  let statement: string;
  let notes: string[];
  let constraintMetas: ConstraintMeta[] = [];

  if (resolved.TABLE_TYPE === 'BASE TABLE') {
    const [row] = await exec<Record<string, unknown>>(`SHOW CREATE TABLE ${qualified}`, []);
    const raw = row?.['Create Table'];
    if (typeof raw !== 'string' || raw.length === 0) {
      throw new AdapterError(
        'E_QUERY',
        `SHOW CREATE TABLE returned no definition for "${database}"."${object.name}"`,
      );
    }
    statement = stripOneTrailingSemicolon(raw);
    notes = ['Triggers and grants are not included in SHOW CREATE TABLE.'];
    constraintMetas = await listConstraints(exec, database, object.name);
  } else if (resolved.TABLE_TYPE === 'VIEW') {
    const [row] = await exec<Record<string, unknown>>(`SHOW CREATE VIEW ${qualified}`, []);
    const raw = row?.['Create View'];
    if (typeof raw !== 'string' || raw.length === 0) {
      throw new AdapterError(
        'E_QUERY',
        `SHOW CREATE VIEW returned no definition for "${database}"."${object.name}"`,
      );
    }
    statement = stripOneTrailingSemicolon(raw);
    notes = [
      "This is the server's own SHOW CREATE VIEW text, including its DEFINER and SQL SECURITY clauses.",
    ];
  } else {
    throw new AdapterError(
      'E_UNSUPPORTED',
      `definition is not supported for ${resolved.TABLE_TYPE}`,
    );
  }

  return {
    path: encodePath(segments),
    kind: object.kind,
    qualifiedName: `${database}.${object.name}`,
    statements: [statement],
    language: 'sql',
    origin: 'server',
    notes,
    constraints: constraintMetas,
    documentSchema: null,
    sections: [],
    generatedAt: new Date().toISOString(),
  };
}

File: shell/internal/adapters/mysqlfamily/definition.go
package mysqlfamily

import (
	"context"
	"database/sql"
	"time"

	"github.com/kirathecat/kira-studio/shell/internal/adapters"
	"github.com/kirathecat/kira-studio/shell/internal/storage/model"
)

var constraintTypeName = map[string]string{
	"PRIMARY KEY": "primaryKey", "UNIQUE": "unique", "FOREIGN KEY": "foreignKey", "CHECK": "check",
}

// listConstraints is definition.ts's listConstraints. MariaDB/MySQL have no
// pg_get_constraintdef-style builtin (P19 D11) — the key-column list and the FK's referenced
// table/columns are composed from information_schema itself.
func listConstraints(ctx context.Context, exec queryExec, database, table string) ([]model.ConstraintMeta, error) {
	type constraintRow struct {
		name, ctype string
		checkClause *string
	}
	var constraints []constraintRow
	err := exec(ctx, `SELECT tc.CONSTRAINT_NAME AS name, tc.CONSTRAINT_TYPE AS type, cc.CHECK_CLAUSE AS check_clause
	 FROM information_schema.TABLE_CONSTRAINTS tc
	 LEFT JOIN information_schema.CHECK_CONSTRAINTS cc
	   ON cc.CONSTRAINT_SCHEMA = tc.CONSTRAINT_SCHEMA AND cc.CONSTRAINT_NAME = tc.CONSTRAINT_NAME
	 WHERE tc.TABLE_SCHEMA = ? AND tc.TABLE_NAME = ?
	 ORDER BY FIELD(tc.CONSTRAINT_TYPE, 'PRIMARY KEY', 'UNIQUE', 'CHECK', 'FOREIGN KEY'),
	          tc.CONSTRAINT_NAME`, []any{database, table}, func(rows *sql.Rows) error {
		var r constraintRow
		if err := rows.Scan(&r.name, &r.ctype, &r.checkClause); err != nil {
			return err
		}
		constraints = append(constraints, r)
		return nil
	})
	if err != nil {
		return nil, err
	}
	if len(constraints) == 0 {
		return nil, nil
	}

	type keyColumnRow struct {
		name, col        string
		refTable, refCol *string
	}
	var keyColumns []keyColumnRow
	err = exec(ctx, `SELECT CONSTRAINT_NAME AS name, COLUMN_NAME AS col,
	        REFERENCED_TABLE_NAME AS ref_table, REFERENCED_COLUMN_NAME AS ref_col
	 FROM information_schema.KEY_COLUMN_USAGE
	 WHERE TABLE_SCHEMA = ? AND TABLE_NAME = ?
	 ORDER BY CONSTRAINT_NAME, ORDINAL_POSITION`, []any{database, table}, func(rows *sql.Rows) error {
		var r keyColumnRow
		if err := rows.Scan(&r.name, &r.col, &r.refTable, &r.refCol); err != nil {
			return err
		}
		keyColumns = append(keyColumns, r)
		return nil
	})
	if err != nil {
		return nil, err
	}
	columnsByConstraint := map[string][]keyColumnRow{}
	for _, r := range keyColumns {
		columnsByConstraint[r.name] = append(columnsByConstraint[r.name], r)
	}

	metas := make([]model.ConstraintMeta, len(constraints))
	for i, c := range constraints {
		typ, ok := constraintTypeName[c.ctype]
		if !ok {
			typ = "check"
		}
		if typ == "check" {
			def := ""
			if c.checkClause != nil {
				def = *c.checkClause
			}
			metas[i] = model.ConstraintMeta{Name: c.name, Type: typ, Definition: def}
			continue
		}
		cols := columnsByConstraint[c.name]
		colNames := make([]string, len(cols))
		for j, r := range cols {
			colNames[j] = r.col
		}
		columnList := "(" + joinComma(colNames) + ")"
		if typ == "foreignKey" {
			refTable := ""
			if len(cols) > 0 && cols[0].refTable != nil {
				refTable = *cols[0].refTable
			}
			refCols := make([]string, len(cols))
			for j, r := range cols {
				if r.refCol != nil {
					refCols[j] = *r.refCol
				}
			}
			metas[i] = model.ConstraintMeta{
				Name: c.name, Type: typ,
				Definition: columnList + " REFERENCES " + refTable + " (" + joinComma(refCols) + ")",
			}
			continue
		}
		metas[i] = model.ConstraintMeta{Name: c.name, Type: typ, Definition: columnList}
	}
	return metas, nil
}

func joinComma(parts []string) string {
	out := ""
	for i, p := range parts {
		if i > 0 {
			out += ", "
		}
		out += p
	}
	return out
}

// buildDefinition is definition.ts's buildDefinition: passes SHOW CREATE … through verbatim —
// MariaDB/MySQL are asked, never composed.
func buildDefinition(ctx context.Context, exec queryExec, segments []model.PathSegment, database string, objectKind, objectName string) (model.ObjectDefinition, error) {
	var tableType string
	found := false
	err := exec(ctx, `SELECT TABLE_TYPE FROM information_schema.TABLES WHERE TABLE_SCHEMA = ? AND TABLE_NAME = ?`,
		[]any{database, objectName}, func(rows *sql.Rows) error {
			found = true
			return rows.Scan(&tableType)
		})
	if err != nil {
		return model.ObjectDefinition{}, err
	}
	if !found {
		return model.ObjectDefinition{}, adapters.New(adapters.CodeNotFound, "relation \""+database+"\".\""+objectName+"\" not found", nil)
	}

	qualified := quoteIdent(database) + "." + quoteIdent(objectName)
	var statement string
	var notes []string
	var constraintMetas []model.ConstraintMeta

	switch tableType {
	case "BASE TABLE":
		raw, err := showCreate(ctx, exec, "SHOW CREATE TABLE "+qualified, "Create Table")
		if err != nil {
			return model.ObjectDefinition{}, err
		}
		if raw == "" {
			return model.ObjectDefinition{}, adapters.New(adapters.CodeQuery, "SHOW CREATE TABLE returned no definition for \""+database+"\".\""+objectName+"\"", nil)
		}
		statement = adapters.StripOneTrailingSemicolon(raw)
		notes = []string{"Triggers and grants are not included in SHOW CREATE TABLE."}
		constraintMetas, err = listConstraints(ctx, exec, database, objectName)
		if err != nil {
			return model.ObjectDefinition{}, err
		}
	case "VIEW":
		raw, err := showCreate(ctx, exec, "SHOW CREATE VIEW "+qualified, "Create View")
		if err != nil {
			return model.ObjectDefinition{}, err
		}
		if raw == "" {
			return model.ObjectDefinition{}, adapters.New(adapters.CodeQuery, "SHOW CREATE VIEW returned no definition for \""+database+"\".\""+objectName+"\"", nil)
		}
		statement = adapters.StripOneTrailingSemicolon(raw)
		notes = []string{"This is the server's own SHOW CREATE VIEW text, including its DEFINER and SQL SECURITY clauses."}
	default:
		return model.ObjectDefinition{}, adapters.New(adapters.CodeUnsupported, "definition is not supported for "+tableType, nil)
	}

	return model.ObjectDefinition{
		Path: model.EncodePath(segments), Kind: objectKind, QualifiedName: database + "." + objectName,
		Statements: []string{statement}, Language: "sql", Origin: "server", Notes: notes,
		Constraints: constraintMetas, DocumentSchema: nil, Sections: []model.DefinitionSection{},
		GeneratedAt: time.Now().UTC().Format(time.RFC3339Nano),
	}, nil
}

// showCreate runs a SHOW CREATE … statement and returns its named result column (e.g. "Create
// Table") as a string — go's database/sql scans by position, so this reads every column into
// sql.RawBytes and picks the one whose name matches.
func showCreate(ctx context.Context, exec queryExec, statement, columnName string) (string, error) {
	var out string
	err := exec(ctx, statement, nil, func(rows *sql.Rows) error {
		cols, err := rows.Columns()
		if err != nil {
			return err
		}
		raw := make([]sql.RawBytes, len(cols))
		dest := make([]any, len(cols))
		for i := range raw {
			dest[i] = &raw[i]
		}
		if err := rows.Scan(dest...); err != nil {
			return err
		}
		for i, c := range cols {
			if c == columnName {
				out = string(raw[i])
			}
		}
		return nil
	})
	return out, err
}
